hashicorp/terraform
# Cloud run tasks (post-plan only) CLI integration

Integration with Terraform Cloud Run Tasks (post-plan stage). Run tasks is a beta feature implemented on Terraform Cloud that will display results after a plan like so:

<img width="1423" alt="Screen Shot 2021-12-10 at 11 30 52 AM" src="https://user-images.githubusercontent.com/21225410/145631010-473234b2-f443-4ded-8f45-cad39b7d030a.png">

These results are currently not being displayed or consumed by Terraform CLI, so with this PR (as well as some changes made to go-tfe https://github.com/hashicorp/go-tfe/pull/286), is implementing the necessary changes to display the output of Run Tasks' result in the CLI:

<img width="864" alt="Screen Shot 2021-12-10 at 11 14 50 AM" src="https://user-images.githubusercontent.com/21225410/145631442-888a0769-0426-4260-b30f-1441023cc39f.png">

### IMPLEMENTATION

All changes in this PR are being applied through the `op.plan` because we want the user to see results of Run Tasks after a plan has found changes to infrastructure and right before the apply operation (so what we call the `pre-apply` stage).

These changes are also scope to the `cloud` package. This is to be determined, but we think that the CLI will not offer support for Run Tasks prior to version 1.2

This PR also adds an abstraction when interacting with Cloud Integrations such as Run Tasks, Policy Checks and Cost estimation. The abstraction is not yet applied to Policy Checks and Cost Estimation. But if we were to use it for all three Cloud Integrations, it would simplify and unify output during the pre-apply stage.

### TESTING

1. Ensure your organization has access to the Run Tasks beta. You may have to request access for staging and prod. However, if you are running Atlas on dev mode, you'll have Run Task feature enabled by default.
2. Follow these steps to add run tasks to the workspace you'll be testing with: https://www.terraform.io/docs/cloud/workspaces/run-tasks.html
3. We have a go server running on a EC2 instance available for test event hooks. This are the URLs you can add to your run tasks:

http://54.167.177.151/success
http://54.167.177.151/failed
http://54.167.177.151/success?timeout=630 to force a timeout. But you can change the value of timeout to as many seconds you want the server to wait before firing off the webhook back to TFC

4. Create some config using a `cloud` config

```
terraform {
  cloud {
    hostname = "app.staging.terraform.io"
    organization = "test"
    workspaces {
      name = "project"
    }
  }
}

variable "replicas" {
  type = number
  default = 7
}

output "hello_there" {
  value = var.replicas
}

```

5. Using a testing workspace, run `./terraform apply` to see Run Tasks output right after Cost Estimation output. If you have no tasks added to your workspace in the platform, you should see no output related to Run Tasks.

### Screenshots

Task results
![Screen Shot 2021-12-15 at 9 53 53 AM](https://user-images.githubusercontent.com/174332/146238394-8dde1f58-3380-4198-85a3-987e7f314b84.png)

No run tasks defined
![Screen Shot 2021-12-15 at 9 59 56 AM](https://user-images.githubusercontent.com/174332/146238350-03c5ae9e-74ed-470f-a717-f0fa7a2ca4bd.png)

No infrastructure changes (unreachable tasks)
![Screen Shot 2021-12-15 at 10 00 05 AM](https://user-images.githubusercontent.com/174332/146238299-ad5bd407-0204-4fe2-928f-b5047bf9a98b.png)

NOTE: Once this PR has being approved, we will squash the commits.
